Recursive file search with in-content search, closes #1295, by @adelobosko #1337

adelobosko opened the pull request:
https://github.com/gsantner/markor/issues/1295

adelobosko commented:
https://user-images.githubusercontent.com/56083544/118863265-5b152280-b8e7-11eb-989a-023e7ddfd408.mp4

adelobosko commented:
https://user-images.githubusercontent.com/56083544/118864668-e0e59d80-b8e8-11eb-9850-4484bc86ad20.mp4

gsantner commented:
I did various improvements all over the plcae. Also ~ halfed the amount of changed files, which makes it _less hard_ to review.

For that, already transitioned some generic code parts to master branch.

gsantner commented:
I did various improvements all over the placed. Also ~ halved the amount of changed files, which makes it _less hard_ to review.

For that, already transitioned some generic code parts to master branch.

gsantner commented:
@harshad1  
If you are interested, please feel free to help review.  
It's just a quite massive PR on many areas. It's already open quite some time, and shelfed off many things already - but still quite big.

harshad1 commented on app/src/main/java/net/gsantner/markor/activity/DocumentActivity.java:
```diff
@@ -66,13 +67,17 @@
 
-    public static void launch(Activity activity, File path, Boolean isFolder, Boolean doPreview, Intent intent) {
+
+    public static void launch(Activity activity, File path, Boolean isFolder, Boolean doPreview, Intent intent, final int... arg0lineNumber) {
```
It seems the variadic `final int... arg0lineNumber` is not used for multiple values, but as a nullable integer. If this is the case, I suggest replacing it with `Integer lineNumber`, which can also be set to null

gsantner replied:
changed

harshad1 commented on app/src/main/java/net/gsantner/markor/activity/DocumentActivity.java:
```diff
@@ -271,14 +278,10 @@ public void setDocumentTitle(final String title) {
-
-    public void showTextEditor(@Nullable Document document, @Nullable File file, boolean fileIsFolder, boolean preview) {
-
+    public void showTextEditor(@Nullable Document document, @Nullable File file, boolean fileIsFolder, boolean preview, final int... arg0fileLineNumber) {
```
Same observation about using an overload. Or just use `Integer lineNumber` and pass in null wherever needed

harshad1 replied:
@adelobosko missed this one? :)

harshad1 commented on app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java:
```diff
@@ -491,7 +482,7 @@ public boolean onOptionsItemSelected(final MenuItem item) {
                 FilesystemViewerCreator.showFileDialog(new FilesystemViewerData.SelectionListenerAdapter() {
                                                            @Override
-                                                           public void onFsViewerSelected(String request, File file) {
+                                                           public void onFsViewerSelected(String request, File file, final Integer... lineNumber) {
```
variadic as optional

harshad1 commented on app/src/main/java/net/gsantner/markor/activity/SettingsActivity.java:
```diff
@@ -264,7 +264,7 @@ public Boolean onPreferenceClicked(Preference preference, String key, int keyRes
                         FilesystemViewerCreator.showFolderDialog(new FilesystemViewerData.SelectionListenerAdapter() {
                             @Override
-                            public void onFsViewerSelected(String request, File file) {
+                            public void onFsViewerSelected(String request, File file, final Integer... lineNumber) {
```
Variadic as optional

harshad1 commented on app/src/main/java/net/gsantner/markor/util/AppSettings.java:
```diff
@@ -592,6 +609,82 @@ public boolean isMultiWindowEnabled() {
     }
 
+    public boolean isShowSearchResultOnCancel() {
+        return getBool(R.string.pref_key__is_show_search_result_on_cancel_enabled, true);
```
Note: We can use backup-restore to change these values (if necessary for testing etc)

harshad1 commented:
Still reviewing, will take a couple days for me given my time constraints, but the code looks pretty good so far. Only issue I have seen is this pattern of using a variadic (Varargs) as an optional arg. This isn't a great approach, IMO, but feel free to convince me otherwise.

gsantner commented:
Yeah the variadic actually are from me by PR changes. This way its optional to specify for callback

adelobosko commented:
>  I prefer only one single streamedline.
i have changed Integer params to one Integer parameter, I think it will be better.

harshad1 commented:
> Yeah the variadic actually are from me by PR changes. This way its optional to specify for callback.
> 
> There were previously 2 callback methods for selected and I prefer only one single streamedline.

Then the Integer method is the way to go, IMO. Using variadic is really confusing unless one studies the code underneath. Personally, I am much happier with the changes @adelobosko just made

harshad1 commented on app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java:
```diff
@@ -680,18 +666,54 @@ public boolean onBackPressed() {
+    }
+
+    public void restoreDocumentPositions() {
+        if (!Arrays.asList(_hlEditor, _webView, _document.getFile(), _primaryScrollView).contains(null) && !isTodoOrQuickNote()) {
```
Neat trick!

gsantner replied:
> may not work if all are null.

why? [null, null, null].contains(null) => true

harshad1 commented on app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java:
```diff
@@ -377,14 +360,17 @@ public boolean onOptionsItemSelected(final MenuItem item) {
             }
             case R.id.action_preview: {
+                saveDocumentPositions();
                 setDocumentViewVisibility(true);
```
Does it make more sense to call `saveDocumentPositions` inside `setDocumentViewVisibility`

gsantner replied:
might be that setting cursor happens after saving (=->always 0), caution

adelobosko replied:
> 
> 
> might be that setting cursor happens after saving (=->always 0), caution

yes, it is a reason that I do not move `saveDocumentPositions` into `setDocumentVisibility`

harshad1 commented on app/src/main/java/net/gsantner/opoc/ui/FileSearchDialog.java:
```diff
@@ -0,0 +1,224 @@
+        searchEditText.setSingleLine(true);
+        searchEditText.setMaxLines(1);
+        searchEditText.setTextColor(textColor);
+        searchEditText.setHintTextColor((textColor & 0x00FFFFFF) | 0x99000000);
```
Should make the color a resource

gsantner replied:
It is a color filter applied to a existing color. should be fine I'd say

harshad1 replied:
👍

harshad1 commented on app/src/main/java/net/gsantner/markor/ui/SearchOrCustomTextDialogCreator.java:
```diff
@@ -120,26 +123,33 @@ public static void showInsertTableRowDialog(final Activity activity, final boole
+            appSettings.setSearchQueryCaseSensitivity(cb_options._isCaseSensitiveQuery);
+            appSettings.setSearchInContent(cb_options._isSearchInContent);
+            appSettings.setOnlyFirstContentMatch(cb_options._isOnlyFirstContentMatch);
+            SearchEngine.Config config = new SearchEngine.Config(searchDir, cb_options._query, isShowResultOnCancel, maxSearchDepth, ignoredDirs, ignoredFiles, cb_options._isRegexQuery, cb_options._isCaseSensitiveQuery, cb_options._isSearchInContent, cb_options._isOnlyFirstContentMatch, isShowMatchPreview, searchByContentExtensions);
```
Would it be possible to use a dopt like mechanism here? These long function calls are hard to follow

adelobosko replied:
> 
> Looks like I do not have push rights. Code is ready though. @adelobosko could you give me permission?

Sure, I sent you the invite

harshad1 commented on app/src/main/java/net/gsantner/markor/ui/SearchOrCustomTextDialogCreator.java:
```diff
@@ -120,26 +123,33 @@ public static void showInsertTableRowDialog(final Activity activity, final boole
+            final List<String> ignoredFiles = appSettings.getIgnoredSearchFileNames();
+            final boolean isShowMatchPreview = appSettings.isShowMatchPreview();
+            final List<String> searchByContentExtensions = appSettings.getSearchByContentExtensions();
+            appSettings.setSearchQueryRegexUsing(cb_options._isRegexQuery);
```
Do we really want to remember these options for next time? @gsantner

gsantner replied:
I think we should carefully think what all settings we want to have. however putting PR review focus atm on the funcationalties and not breaking existing code

harshad1 replied:
:+1

harshad1 commented on app/src/main/java/net/gsantner/opoc/ui/SearchEngine.java:
```diff
@@ -0,0 +1,570 @@
+
+
+        @Override
+        protected void onProgressUpdate(Integer... values) {
```
Looks like we always expect exactly 4 args. Should just name them.

gsantner replied:
this is a method from Android AsyncTask class. Signature should match

harshad1 replied:
Ah sorry, nevermind

harshad1 commented:
Right now, this searches through PDF, JPG and other binary files. We should add logic to test if a file is a text file before searching it.

gsantner commented:
This is a common issue where we dont have a generic solution yet. But please not lets implement this in the PR.

Instead, lets have a simple if check to not read files bigger than 800kb. For example.

harshad1 commented:
> This is a common issue where we dont have a generic solution yet. But please not lets implement this in the PR.

We have `Files.probeContentType` and `URLConnection.guessContentTypeFromStream` built into java. Additionally we already have the `TextFormat.isTextFile` routine. 

I can add a function to do this test as a separate PR. We can then use it here. Should have it up in a few hours
    
> Instead, lets have a simple if check to not read files bigger than 800kb. For example.

I don't think this is a very good heuristic. Most of the PDFs and images I am getting in my search are < 800 kb. They really pollute the search results and increase the search time.

gsantner commented:
> We have Files.probeContentType and URLConnection.guessContentTypeFromStream built into java. Additionally we already have the TextFormat.isTextFile routine.

yeah we can try. however mentioned java methods often do not work as expected, or are not supported on older android(=java) versions.

> I can add a function to do this test as a separate PR. We can then use it here. Should have it up in a few hours

ehm I think @adelobosko  already gave you push rights to his repository. can you try this way, push directly to his PR?

harshad1 commented:
> ehm I think @adelobosko already gave you push rights to his repository. can you try this way, push directly to his PR branch?

Looks like I do not have push rights. Code is ready though. @adelobosko could you give me permission?

harshad1 commented on app/src/main/res/xml/preferences_master.xml:
```diff
@@ -71,6 +71,50 @@
+                android:title="@string/ignored_search_file_names" />
+            <android.support.v7.preference.EditTextPreference
+                android:defaultValue="@string/empty_string"
+                android:dialogMessage="@string/search_by_context_have_to_include_dialog"
```
Instead of having this and the previous setting, I think it may be better to just have a filename filter directly in the dialog. 

Default filter file would be `*.*`, and a user would be able to enter, say, `*.md`.

adelobosko replied:
Yes, it is more convenient

harshad1 commented on app/src/main/java/net/gsantner/opoc/ui/SearchEngine.java:
```diff
@@ -0,0 +1,570 @@
+            Integer queueDepth = values[1];
+            Integer filesFound = values[2];
+            Integer countCheckedFiles = values[3];
+            String snackBarText = "f:" + filesFound + " qu:" + queueLength + "|" + queueDepth + " c:" + countCheckedFiles + "\n" + _config._query;
```
Not very important for this PR, but we should consider replacing the letters with icons in the future.

harshad1 commented:
@adelobosko Could you give me push permissions for your branch? I have more improvements I would like to make to filtering and it would be easier to push rather than to create PRs

harshad1 commented:
> see opoc utils, there is a similar but better working version already...with text/markdown for example

Sorry, what is the context for this?

gsantner commented:
![grafik](https://user-images.githubusercontent.com/6735650/119275974-700bf180-bc18-11eb-9984-6a5eae952c2a.png)

harshad1 commented on app/src/main/java/net/gsantner/opoc/ui/FileSearchResultSelectorDialog.java:
```diff
@@ -0,0 +1,247 @@
+    }
+
+
+    private static Pair<ArrayList<Map<String, Object>>, ArrayList<ArrayList<Map<String, Object>>>> filter(final List<SearchEngine.FitFile> searchResults, String query) {
```
@adelobosko If you are able to document this data structure, that would be helpful for future maintainece.

adelobosko commented:
> @adelobosko Could you give me push permissions for your branch? I have more improvements / bug fixes I would like to make to filtering and it would be easier to push rather than to create PRs

Hmm, i sent in ite to you yesterday, couldnyou check?

harshad1 commented:
> Hmm, i sent in ite to you yesterday, couldnyou check?

![image](https://user-images.githubusercontent.com/1940940/119385763-767f8500-bc7b-11eb-834d-196fede7654a.png)

I believe you'll have to add me as a collaborator:

https://docs.github.com/en/github/setting-up-and-managing-your-github-user-account/managing-access-to-your-personal-repositories/inviting-collaborators-to-a-personal-repository

adelobosko commented:
> ![image](https://user-images.githubusercontent.com/1940940/119385763-767f8500-bc7b-11eb-834d-196fede7654a.png)
> 
> I believe you'll have to add me as a collaborator:

I have done it. 
maybe this link will help you:
https://github.com/adelobosko/markor/invitations

![image](https://user-images.githubusercontent.com/56083544/119389482-5d56f280-bcd4-11eb-9a3a-f4a29352ceb0.png)

harshad1 commented:
@adelobosko Thanks! got it.

harshad1 commented:
IMO one more thing we do need to add is shortening very long lines in the search results.

Perhaps something like  `. . . < first match in line > . . . `

harshad1 commented:
> IMO one more thing we do need to add is shortening very long lines in the search results.

I have added support for this.

@adelobosko  @gsantner IMO this is ready to be merged.

There will always be things which can be tweaked, but I think they should be discussed and made in subsequent PRs, if needed.

harshad1 commented:
Actually, one more thing: I am noticing that the keyboard no longer opens when I open a file. Was this change made intentionally?

adelobosko commented:
> Was this change made intentionally?
Yes, need to tweak it. Let me do this

harshad1 commented:
> Yes, need to tweak it. Let me do this

👍

harshad1 commented:
@adelobosko are your changes ready for me to test?

adelobosko commented:
> @adelobosko are your changes ready for me to test?

@harshad1 yes, sorry that did not inform you

harshad1 commented:
@adelobosko Not working consistently. I am working on ways to consistently show the keyboard after the animation (using AnimatorListener), but none seem to work

It may be best to just use setSelection for now and come back to this in a later PR. Will push my changes tomorrow, if they are successful

adelobosko commented:
I think just need to increase post delay in the showSoftKeyboard()

adelobosko commented:
@harshad1 could you try it again?

gsantner commented:
> animation (using AnimatorListener), but none seem to work

sure that the keyboard call happens on UI thread?

harshad1 commented:
> @harshad1 could you try it again?

Keyboard now opens, but the following issues remain:

1. Keyboard remains open in file explorer view. To reproduce
    1. Open document
    1. press back arrow at the top of the document
2. Quicknote and todo.txt have no selection set (and keyboard does not open). They are supposed default jump to the end of the file.
3. If I switch to another app and then back to the document, the selection and keyboard are not restored

I strongly feel we should restore all the logic in `DocumentEditFragment`, except for the minimum needed to support launching search intents, and address these changes it in a separate PR. There are a lot of nuances here related to activity lifecycle.

DarkP0wer commented:
> Keyboard remains open in file explorer view. To reproduce 
Maybe remove the smooth selection / delays when open, and just select a position?

>     2. Quicknote and todo.txt have no selection set (and keyboard does not open). They are supposed default jump to the end of the file.

This is not supported in the current PlayMarket's version

>     3. If I switch to another app and then back to the document, the selection and keyboard are not restored

and this as well, so it is not a bug from this PR and we can add it in the future with the next pull.

adelobosko commented:
> Keyboard remains open in file explorer view. To reproduce 
Maybe remove the smooth selection / delays when open, and just select a position?

>     2. Quicknote and todo.txt have no selection set (and keyboard does not open). They are supposed default jump to the end of the file.

This is not supported in the current PlayMarket's version

>     3. If I switch to another app and then back to the document, the selection and keyboard are not restored

and this as well, so it is not a bug from this PR and we can add it in the future with the next pull.

harshad1 commented:
> This is not supported in the current PlayMarket's version, so it is not a bug from this PR and we can add it in the future with the next pull.

Hmmm. I am using markor from F-Droid and it works fine is the play store version that different. To be clear, I mean Quicknote and todo.txt as standalone files, not in the tabs. (Create by long pressing the markor icon -> dragging todo.txt).

On my device, both these files jump to the bottom.

gsantner commented:
play store may lag behind in releases, but generally speaking it's 99.99% same code.

adelobosko commented:
> To be clear, I mean Quicknote and todo.txt as standalone files, not in the tabs.

Hmm yes, I will fix it

gsantner commented:
yeah please lets tidy the PR as far the feature is concered. but not lets add general/unrelated fixes to it

harshad1 commented on app/src/main/java/net/gsantner/markor/activity/MainActivity.java:
```diff
@@ -409,9 +409,9 @@ public void onFsViewerDoUiUpdate(FilesystemViewerAdapter adapter) {
+                public void onFsViewerSelected(String request, File file, final Integer lineNumber) {
                     if (TextFormat.isTextFile(file)) {
-                        DocumentActivity.launch(MainActivity.this, file, false, null, null);
+                        DocumentActivity.launch(MainActivity.this, file, false, null, null, lineNumber);
```
@adelobosko  Is `lineNumber` actually used / passed in anywhere in `onFsViewerSelected`? If not, we should revert the signature and pass in null here?

adelobosko replied:
> 
> 
> @adelobosko Is `lineNumber` actually used / passed in anywhere in `onFsViewerSelected`? If not, we should revert the signature and pass in null here?

Sorry, I do not understand the question.

adelobosko replied:
> 
> 
> @adelobosko Is `lineNumber` actually used / passed in anywhere in `onFsViewerSelected`? If not, we should revert the signature and pass in null here?

if we replace lineNumber here to null then we could not move to this line in the file
![image](https://user-images.githubusercontent.com/56083544/119893234-c8f2c700-bf43-11eb-852b-d892bc7a380b.png)

adelobosko commented:
> yeah please lets tidy the PR as far the feature is concered. but not lets add general/unrelated fixes to it

ok, I can sort it tomorrow or at next day after tomorrow

harshad1 commented on app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java:
```diff
@@ -680,18 +663,75 @@ public boolean onBackPressed() {
+                                _hlEditor.requestFocus();
+                            }
+                            _hlEditor.setSelection(selPosition);
+                            new ActivityUtils(getActivity()).showSoftKeyboard().freeContextRef();
```
So far, i have not found any issues in my testing this latest version. Only some cleanup needed if I can't find more behavior bugs. :+1:

harshad1 replied:
@adelobosko I do not think that this postDelayed is actually necessary. `restoreDocumentPositions` is called in `onResume`. UI actions should be fine there without any delay. Works fine on my device without `postDelayed`

adelobosko replied:
> 
> 
> @adelobosko I do not think that this postDelayed is actually necessary. `restoreDocumentPositions` is called in `onResume`. UI actions should be fine there without any delay. Works fine on my device without `postDelayed`

@harshad1 hmm, for me delay is needed. I tried without the delay but the selection could not set. Try to open a big file. I think need another way to set selection. After the editor text is loading... i going to remove restore selection in this pull

adelobosko replied:
> After the editor text is loading...

But seems 300 milliseconds is enough for loading a big file (selection is set). Need to try a very big file :)

harshad1 commented on app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java:
```diff
@@ -680,18 +663,75 @@ public boolean onBackPressed() {
+        new ActivityUtils(getActivity()).hideSoftKeyboard().freeContextRef();
+        _hlEditor.postDelayed(() -> {
+            _hlEditor.clearFocus();
+            new ActivityUtils(getActivity()).hideSoftKeyboard().freeContextRef();
```
@adelobosko Same comment about postDelayed may not be necessary

adelobosko reviewed (commented):
i think it is ready

gsantner commented:
@adelobosko 

Why did you push `reverted the "start editor on bottom" option`? Intentionally did the change, as it works slightly different now.

adelobosko commented:
> @adelobosko 
> 
> Why did you push `reverted the "start editor on bottom" option`? Intentionally did the change, as it works slightly different now, and all clients have it enabled currently as it's the old default.

@gsantner 
Because I thought that is related to the "restore position" commit that was reverted before...

gsantner commented:
@adelobosko 

Just one "not" filename filter should be fine. I tend to suggestion of combining into one. This is also what I'm used to for example at Syncthing <https://docs.syncthing.net/users/ignoring.html#example>. Having two options is overkill.

![screenshot-2021-05-31__21-03-15](https://user-images.githubusercontent.com/6735650/120232356-b504d900-c253-11eb-9451-f591f91cfdd5.png)

Do you agree or is there a bigger issue at implementation with that? I mean `.*\.mp4` can be applied folderpaths `/storage/emulated/0/video.mp4` just in the same manner.

adelobosko commented:
@gsantner so do I need to discard my revert :) (`start on the bottom`)?

gsantner commented:
Well I wanted to know why you reverted back to the original key :D

adelobosko commented:
> I mean `.*\.mp4`
> 

Yes, it will be fine

gsantner commented:
I'm sorry but I'm not happy with multiple signatures and callback methods, carrying usually the same result ("finished search").

Trying to make efforts for instead: Callback.a1<String, Hashmap>, with the second one containing key value pairs of additional metadata.


![grafik](https://user-images.githubusercontent.com/6735650/120233134-68ba9880-c255-11eb-852c-7da6e8574ab7.png)


![grafik](https://user-images.githubusercontent.com/6735650/120233121-61938a80-c255-11eb-9abf-83e775b50eea.png)

adelobosko commented:
Not clear to me, do i need back the revert of `start on the bottom commit`?

I reverted `the restore position commit` because i want to create a new pull with this.
And I reverted `the start on bottom commit`, because I thought that is related to `restore the position commit`

gsantner commented:
I had the finding that, by at least changes of the PR - start on bottom and recent position restore didn't work as expected. But with the default configuration that was in the PR it worked better for me

gsantner commented:
@adelobosko 

unused method, can be removed right?

![grafik](https://user-images.githubusercontent.com/6735650/120239075-cc48c400-c25d-11eb-8ac0-c8112aa4a573.png)

adelobosko commented:
@harshad1 getIndexFromLineOffset working bad... please move my function back...

adelobosko commented:
@harshad1 for example:
String t1 = "2123";
int index = getIndexFromLineOffset(t1, 0, 0);
and sometimes it moves to the end of the previous line but not to start...

adelobosko commented:
> 
> 
> @harshad1 for example:
> String t1 = "2123";
> int index = getIndexFromLineOffset(t1, 0, 0); //index = 4 not 0
> and sometimes it moves to the end of the previous line but not to start...

adelobosko commented:
@harshad1 and try to check big files what will be faster...
```java
int index = getIndexFromLineOffset(t1, 100, 0);
```
or
```java
int index = getIndexByLineNumber(t1, 100);
```
getIndexByLineNumber will be faster 4 times as does not have additional calculations, just counts the `\n` char...

adelobosko commented:
> unused method, can be removed right?
@gsantner sure, (it used for a history listview, now it is a spinner)

gsantner commented:
* guess it's ok to get `getIndexByLineNumber` back
* about "ignored dir/file" - should I look into it or have you already started adelobosko?

harshad1 commented:
@adelobosko @gsantner I am investigating the performance and behavior difference, I will revert if I cannot fix `getIndexFromLineOffset`

harshad1 commented:
@adelobosko Note that `getIndexFromLineOffset(foo, n, 0)` will move cursor to the _end_ of the nth line, not the start. The difference didn't matter for scrolling to the line, so I did not add a `getLineStart` call.

I do not think performance is an issue, `getIndexFromLineOffset` is used a *lot* in code which must be responsive (such as `regexReplaceAction`) and no issues with behavior or performance are observed. Unless you are calling it in a tight loop, the current code will do.

adelobosko commented:
> about "ignored dir/file" - should I look into it or have you already started adelobosko

Ok, sure. I think better that you look at it.

adelobosko commented:
> 
> Note that `getIndexFromLineOffset(foo, n, 0)` will move cursor to the _end_ of the nth line, not the start. 

ok

gsantner commented:
btw, when this is merged I want to release Markor update

gsantner commented:
whats the status on the cursor moving? Fine now or are you both happy with it?

gsantner commented:
whats the status on the cursor moving/line-index? Fine now / are you both happy with it?

harshad1 commented:
> whats the status on the cursor moving/line-index? Fine now / are you both happy with it?

I believe we have agreed do use the old version for now. 

If performance or behavior becomes a concern in the future, I will update the existing function to address it.

adelobosko commented:
> I believe we have agreed do use the old version for now.

yes, i am confirming

harshad1 commented:
Anything I can do to help move this along?

gsantner commented:
@harshad1 
getIndexByLineNumber was wrongly merged ahead, now removed again as we keep the old solution with getIndexFromLineOffset. (Which is now slightly improved).

Is this correct? / So reverted as expected already in the PR?

![grafik](https://user-images.githubusercontent.com/6735650/122472472-ec4be780-cfc0-11eb-9267-2697d7047fc5.png)

harshad1 commented:
@gsantner 
> Is this correct? / So reverted as expected already in the PR?

Yes. The current state in the PR is correct. No revert is necessary.

gsantner commented:
ok thx

I think it's still quite clunky, wanna chop of some more. From functionality remaining is merge of the ignored dir/file settings & functions to a single one

harshad1 commented:
I agree. Let me know if you need me test something

gsantner commented:
yeah thanks

and sory that it not moves fast currently. lots of personal stuff todo to take care off

gsantner commented:
yeah thanks

and sorry that it not moves fast currently. lots of personal stuff to take care off.  
(because I've said on weekend I'll merge)

gsantner commented:
~200-300 lines less again

harshad1 commented on app/src/main/java/net/gsantner/markor/ui/fsearch/FileSearchResultSelectorDialog.java:
```diff
@@ -108,7 +109,7 @@ public void beforeTextChanged(final CharSequence arg0, final int arg1, final int
         expandableListView.setOnChildClickListener((parent, v, groupPosition, childPosition, id) -> {
             GroupItemsInfo groupItem = (GroupItemsInfo) parent.getExpandableListAdapter().getGroup(groupPosition);
-            ChildItemsInfo childItem = (ChildItemsInfo) parent.getExpandableListAdapter().getChild(groupPosition, childPosition);
+            SearchEngine.FitFile.ContentMatchUnit childItem = (SearchEngine.FitFile.ContentMatchUnit) parent.getExpandableListAdapter().getChild(groupPosition, childPosition);
```
Why not just use a `Pair<Integer, String>` in both places?

gsantner replied:
I already thought about it and on my own thought that it's easier to follow if it's explicit with name "linenumber".

However changed this now, and at the end it's not that bad as we don't access it much often.

gsantner commented:
@adelobosko @harshad1   
Thank you very much for programming this. I further slimmed this down throughout the week and I'm gonna merge it now.  
I'm not 100% happy about everything, but it's mature enough for users 😄 . 

Again, thanks! :D
